### DISCUSSION TITLE
differences with connectorx?

### DISCUSSION THREAD

**Discussion by evbo** [pymssql/pymssql#812] [Q&A]
I use `pymssql` without any difficulties, but the login always fails when using something like `connectorx` instead. What are the differences in how `pymssql` initiates a connection than with `connectorx`? Is `pymssql` more careful about handling url encoded usernames and passwords?

```python
import pymssql
conn = pymssql.connect("fehchdev.com", "FEHCH\oper_dev", "b3k@!!1aa", "registrar_home")
cursor = conn.cursor(as_dict=True)
cursor.execute("select * from dbs1.users_home")
row = cursor.fetchone()
print(row)
# works!
```

```python
from urllib import parse
import connectorx as cx

u = parse.quote_plus("FEHCH\oper_dev")
p = parse.quote_plus("b3k@!!1aa")
host = "fehchdev.com"
db = "registrar_home"
uri = f"mssql://{u}:{p}@{host}:1433/{db}"

cx.read_sql(uri, "select * from dbs1.users_home")
```
fails with:
> tiberius::tds::stream::token Login failed for user...

and:
> RuntimeError('Timed out in bb8')

I'm just trying to understand these differences as pymssql is doing such a quality job. What about it causes it to work so well?!

**Comment by termim** [comment]
Please note that in case of `pymssql` you do not use user/password quoting.
May be `connectorx` do not unquote url fragments?
What if feed it user and password without using `quote_plus`?

**Reply by evbo** [reply]
Thanks, but they definitely document on their github to url encode if there's special characters. I get the same errors if I don't. Unfortunately they haven't responded but it does seem like historically there have been issues with quoting special characters. I was posting here in case there is a fundamentally different approach taken in `pymssql`. But after digging through your source code, it seems no special settings are being made by default...

**Comment by termim** [comment]
If they use `FreeTDS` you can try to play with `TDSDUMP` environment variable to see what they really use for user/password.
